fix(coderd): remove rate limits from agent metadata

ammario opened the pull request:
fix(coderd): remove rate limits from agent metadata

Include the full update message in the PubSub notification so that
we don't have to refresh metadata from the DB and can avoid rate
limiting.

mafredri commented on coderd/workspaceagents.go:
```diff
@@ -1567,7 +1564,13 @@ func (api *API) workspaceAgentPostMetadata(rw http.ResponseWriter, r *http.Reque
+		return
+	}
+
+	err = api.Pubsub.Publish(watchWorkspaceAgentMetadataChannel(workspaceAgent.ID), datumJSON)
```
I don't know how limited the pubsub capacity is wrt payload size, but since we don't control the size (of e.g. values), we'll still need a different approach to propagate the value. Not against adding a few more fields to the publish though if they're helpful.

I see a test also caught this.

mafredri commented on coderd/workspaceagents.go:
```diff
@@ -1524,7 +1536,9 @@ func (api *API) workspaceAgentPostMetadata(rw http.ResponseWriter, r *http.Reque
-		maxValueLen = 32 << 10
+		// maxValueLen is set to 4096 to stay under the 8000 byte Postgres
+		// NOTIFY limit.
+		maxValueLen = 4096
```
Can there be both a value and an error, resulting in 2*4096?

ammario replied:
ah yeah

mafredri commented on coderd/workspaceagents.go:
```diff
@@ -1644,34 +1661,26 @@ func (api *API) watchWorkspaceAgentMetadata(rw http.ResponseWriter, r *http.Requ
+			api.Logger.Error(ctx, "failed to unmarshal pubsub message", slog.Error(err))
+			return
 		}
+		pubsubUpdates <- update
```
If `sendMetadata` takes a long time, there's a chance we'll fill up the 8 buffer slots we've got here and then hang.

We could instead append to a slice here and use a buffer of 1 instead to signal that there's new data in the slice to process.

ammario replied:
I think we want to cap the buffer size at something though, at which point we're back to the same problem.

mafredri replied:
Do you mean preventing the slice from growing unboundedly? I suppose we can just limit it's max size and discard the oldest once it has more than X entries? As for notifying we can just do a select with a default on a 1 size chan.

ammario replied:
hmm I can actually just use a map on the key

mafredri commented on coderd/workspaceagents.go:
```diff
@@ -1603,30 +1625,25 @@ func (api *API) watchWorkspaceAgentMetadata(rw http.ResponseWriter, r *http.Requ
 			// We always use the original Request context because it contains
 			// the RBAC actor.
-			lastDBMeta, err = api.Database.GetWorkspaceAgentMetadata(ctx, workspaceAgent.ID)
+			lastMetadata, err = api.Database.GetWorkspaceAgentMetadata(ctx, workspaceAgent.ID)
```
If this fails we should probably disconnect the client. Pull only happens once and without `lastMetadata` populated to all metadata entries, we will never return any info.

mafredri reviewed (commented):
Looks like Postgres isn't liking the Gob (wants UTF8). Other than that and a few other observations, I think this approach is fine. 👍🏻

ammario commented:
2bf75883:

The new approach uses a mutex-protected map to store pending changes and avoids
backpressure problems.

sreya commented on coderd/workspaceagents.go:
```diff
@@ -1567,7 +1568,15 @@ func (api *API) workspaceAgentPostMetadata(rw http.ResponseWriter, r *http.Reque
 	)
 
-	err = api.Pubsub.Publish(watchWorkspaceAgentMetadataChannel(workspaceAgent.ID), []byte(datum.Key))
+	// gob is used instead of JSON due to the potential large performance
+	// overhead of agent metadata.
```
outdated comment

mafredri commented on coderd/workspaceagents.go:
```diff
@@ -1593,7 +1602,42 @@ func (api *API) watchWorkspaceAgentMetadata(rw http.ResponseWriter, r *http.Requ
+	// We avoid channel-based synchronization here to avoid backpressure problems.
+	var (
+		metadataMapMu sync.Mutex
+		metadataMap   = make(map[string]database.WorkspaceAgentMetadatum, 128)
```
Will there always be 128 entries or would it be better to let it grow as needed?

mafredri commented on coderd/workspaceagents.go:
```diff
@@ -1603,97 +1645,57 @@ func (api *API) watchWorkspaceAgentMetadata(rw http.ResponseWriter, r *http.Requ
-	defer refreshTicker.Stop()
+	// We send updates at most every second.
+	const sendInterval = time.Second * 1
+	sendTicker := time.NewTicker(sendInterval)
```
A `NewTimer` might be better, coupled with `Reset` if we want to be literal with the comment above (at most every second). The next ticker could immediately be ready depending on how things went.

ammario replied:
Exactly every second was my intention — updated the comment.

mafredri reviewed (approved):
A few notes but looks good. A `sync.Map` might help clean it up a bit.

ammario commented on coderd/workspaceagents.go:
```diff
@@ -1603,97 +1645,57 @@ func (api *API) watchWorkspaceAgentMetadata(rw http.ResponseWriter, r *http.Requ
 	}
-	defer cancelSub()
+	for _, datum := range md {
+		metadataMap[datum.Key] = datum
```
ah yeah 🤦🏽‍♂️

ammario committed:
fix(coderd): remove rate limits from agent metadata

Include the full update message in the PubSub notification so that
we don't have to refresh metadata from the DB and can avoid rate
limiting.
